pkg/client/restclient/urlprovider.go

```diff
@@ -0,0 +1,84 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package restclient
+
+import (
+	"net/url"
+	"sync"
+)
+
+// A URLProvider allows to use multiple URLs.
+// Also maintains state of currently working URL.
+type URLProvider interface {
+	// Get returns currently selected URL.
+	Get() *url.URL
+	// Next selects any other URL, it is upto implementation to provide any
+	// health checking/load balancing, it might be as simple as an iterator.
+	Next() *url.URL
+	// WrapWithRetry executes function until success or until all urls were visited.
+	WrapWithRetry(func() error) error
+}
+
+// A RoundRobinProvider allows to iterate over URLs, doesnt provide any health checking.
+type RoundRobinProvider struct {
```

Comment:
Pasting from a previous thread: The idea is that if we start to run controllers like scheduler and controller-manager onto the cluster in HA configurations, they should talk to the same API server if possible. Because each apiserver serves from a cache (unless the linearize read flag is passed explicitly) and each apiserver cache could be a different degree behind etcd. This could cause bugs in naive clients that don't handle this. We can avoid these bugs by adding clientIP session affinity.

We should do this but I'm not sure if we need it in the initial implementation.